Automattic/pocket-casts-android
# Settings Sync - Refactor "General" settings to be UserSettings

## Description
This PR is the first step toward refactoring our settings to be `UserSettings` in an effort to make it easier to sync our settings in the future (and easy to have new settings be synced).

I'm only including the settings from the "General" screen in order to get early feedback on this before I invest more time in this direction. Also, this helps make it a bit easier to review the PR (only 32 changed files ! 😅 ).

The main point of this is to simplify what it takes to add a syncable setting so it is easier to add synced settings in the future. For example, with this change, instead of the SettingsImpl class needing to implement four methods/fields:

1. skipForwardInSecsObservable
2. getSkipForwardInSecs
3. setSkipForwardInSecs
4. getSkipForwardNeedsSync

We now just need a single val skipForwardInSecs: UserSetting field, and that also covers the coroutine flow as well (which would be a another field that needed to be implemented before this PR).

> **Note**
> I have this PR targeting a feature branch because I think it would be good to hold these changes back for now. That will allow me to make further changes to how `UserSetting`s work if necessary without having that directly impact our production users. Also, this reserves for us the option of not rolling this change out until we start syncing settings more broadly (I'm not sure that's a good idea, but I'm not sure that it isn't either). Let me know if you think we're better off merging to `main` because I don't feel strongly.

## Testing Instructions
For each setting in the "General" settings:
1. Fresh install the app
2. Verify that the setting has the same default value that it had before
3. Log into an account
4. Change the setting to a new value
5. Force close the app
6. Verify that the setting has retained the new value
7. If the setting is one that is synced (skip-forward and skip-back times):
    1. Sync the app
    2. uninstall the app
    3. reinstall the app and log into the account
    4. Once the account has synced, verify that the updated setting value is shown
8. Verify that the app's behavior reflects the updated setting value (i.e., that skip forward actually skips forward based on the new value).

## Checklist
